docs/proposals/default-service-plans.md

```diff
@@ -0,0 +1,544 @@
+* **{{ service-name }}**: Represents a service where there are not parent-child relationships between service instances. _Parent-child relationships are a new concept in OSB that is still being discussed._ Example: redis, rabbitmq
+
+
+#### Consumer
+
+End users should be able to install a chart on a cluster with any broker that supports this proposal, and not be required to specify any broker-specific values:
+
+
+``` console
+$ helm install azure/wordpress
+# Provision a new service of type mysql and inject it into Wordpress
+```
+
+
+If the cluster operator has created custom classes, that can be specified when installing the chart:
+
+
+``` console
+$ helm install azure/wordpress \
+    --set className=custom-mysql --set planName=ha-mysql
+```
+
+
+
+## Implementation
+
+A key goal of the implementation is to avoid going off and rewriting or requiring a long term branch or fork, etc. I'd like to see this developed using a feature flag, and making all new fields optional.
+
+
+### Service Class
+
+
+| Field | Description |
+|-------|-------------|
+| Service Type | Examples: mysql, redis|
+| <p>Default Provision Parameters</p><p>Default Bind Parameters</p> | <p>Default parameters to supply to the broker during provisioning and binding.</p> <p>OSB API V3 - Make the schema required, and defaults optional.|
+| Default Bind Response Transform | <p>Default json key mapping to transform a non-standard broker response into something that can be relied upon.</p> <p>For example, if a broker returns "database.name" (a nested value) in the response, it can be changed to "database" to un-nest it and apply a standard name.</p>|
+| Controller Reference |<p>Set by Service Catalog when populating the catalog from a broker list to indicate that it is managed by the controller.</p> <p>When not present, the class is ignored during a relist. This gives operators a way to define custom classes.</p>|
+
+Notes
+
+* Relist needs to take into account if the class is managed by the controller or was created by the cluster operator.
+* Defaults can be defined at the Cluster or Namespace level.
+* For simplification, only broker bind response bodies can be transformed, not provision, since that's where we require that functionality today.
+
+
+### Service Plan
+
+| Field | Description |
+|-------|-------------|
+| Service Type | Examples: mysql, redis |
+| annotation: is-default-plan=true | <p>Specifies whether or not a Service Plan is the default for its Service Type.</p> <p>When annotated as the default and a Service Type is set, this plan is used to dynamically provision an instance for that type.</p> |
+| Controller Reference|<p>Set by Service Catalog when populating the catalog from a broker list to indicate that it is managed by the controller.</p> <p>When not present, the plan is ignored during a relist. This gives operators a way to define custom plans.</p> |
+
+
+Notes:
+
+* The controller will enforce that only one default is set per service type.
```

Comment:
What should the controller do when there are two?

Reply:
Good catch! I just realized that I didn't think through how this would work fully. 

Here's how I am thinking default would be defined (below). What do you think?

1. During a list, brokers provide classes/plans. A broker can suggest a plan as a good default by including a tag.
1. Service catalog looks for the service type tag, and also the default tag during list and populates an annotation, "is-suggested-plan" or something along those lines. _I'm not tied to annotations, just going along with what Storage Classes do until we have better ideas_
1. Cluster operators can flag a plan as the default using svcat. This will add a different annotation `is-default-plan`, and Service catalog will handle unsetting the `is-default-plan` annotation on any other plan of the same service type. It doesn't unset the annotation `is-suggested-plan` from the broker, just if the cluster operator has previously picked a different plan as the default.

When service catalog searches for the default plan it will:
1. List all plans that have that are flagged as suggested or default for that service type.
1. When more than one plan has `is-default-plan` set, the operation should fail.
1. When there is one plan with `is-default-plan` set, that plan is used as the default.
1. Otherwise, when more than one plan has `is-suggested-plan` set, the operation should fail. This happens when multiple brokers exist on the cluster, provide a default plan for a particular service type, and the operator has not picked a default.
1. When there is on plan with `is-suggested-plan` set, that plan is used as the default.

In the simplest case, there is one broker that provides default plans, the operator does nothing and people can still get a default plan resolved without any extra work. Otherwise, the operator must step in and pick a default (or the user needs to be specific instead of relying on a default).

Reply:
I've updated the proposal with a how service plan selection works.